services/device-registry-base/src/main/java/org/eclipse/hono/service/management/device/AutoProvisioning.java

```diff
@@ -22,29 +22,122 @@
+     *         <ul>
+     *         <li><em>201 CREATED</em> if the device has successfully been provisioned. The payload contains the
+     *         credentials information.</li>
+     *         <li><em>4XX</em> if the provisioning failed. The payload may contain an error description.</li>
+     *         </ul>
+     * @throws NullPointerException if any of the parameters except clientContext is {@code null}.
+     */
+    public static Future<CredentialsResult<JsonObject>> provisionIfEnabled(
+            final String tenantId,
+            final String authId,
+            final JsonObject clientContext,
+            final CredentialsManagementService credentialsManagementService,
+            final CredentialsService credentialsService,
+            final DeviceManagementService deviceManagementService,
+            final TenantInformationService tenantInformationService,
+            final Span span) {
+
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(authId);
+        Objects.requireNonNull(deviceManagementService);
+        Objects.requireNonNull(credentialsManagementService);
+        Objects.requireNonNull(credentialsService);
+        Objects.requireNonNull(tenantInformationService);
+        Objects.requireNonNull(span);
+
+        return tenantInformationService.getTenant(tenantId, span)
+                .compose(tenantResult -> {
+                    if (tenantResult.isError()) {
+                        return Future.succeededFuture(CredentialsResult.from(tenantResult.getStatus()));
+                    }
+
+                    final TenantObject tenantConfig = tenantResult.getPayload();
+                    return DeviceRegistryUtils
+                            .getCertificateFromClientContext(tenantId, authId, clientContext, span)
+                            .compose(optionalCert -> optionalCert
+                                    .filter(cert -> AutoProvisioning.isEnabled(tenantConfig, cert, span))
+                                    .map(cert -> {
+                                        Tags.ERROR.set(span, Boolean.FALSE); // reset error tag
+                                        return provisionDevice(tenantId, authId, deviceManagementService, credentialsManagementService, cert, span)
+                                                        .compose(r -> {
+                                                            // if error then return error result
+                                                            if (r.isError()) {
+                                                                TracingHelper.logError(span, r.getPayload());
+                                                                return Future.succeededFuture(CredentialsResult
+                                                                        .from(r.getStatus(), new JsonObject()
+                                                                                .put("description", r.getPayload())));
```

Comment:
FMPOV this key should be defined as a constant in e.g. `org.eclipse.hono.utils.Constants` so that we can use it consistently across all of Hono's APIs in the future.